src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/GridGeometryProvider.java

```diff
@@ -0,0 +1,395 @@
+                }
+                return getNativeResolutionGridGeometry();
+            }
+            // Initialize resolution with infinite numbers
+            double[] resolution = new double[] {Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY};
+            ReferencedEnvelope envelope = provider.getBestResolution(features, resolution);
+            AffineTransform at = new AffineTransform(resolution[0], 0, 0, -resolution[1], envelope.getMinX(), envelope.getMaxY());
+            MathTransform tx = ProjectiveTransform.create(at);
+            
+            return new GridGeometry2D(PixelInCell.CELL_CORNER, tx, envelope, GeoTools.getDefaultHints());
+        }
+    }
+
+    /**
+     * Initialize the query to get granules, based on provided filter and region of interest (if any)
+     * 
+     * @param granules
+     * @return
+     * @throws TransformException
+     * @throws FactoryException
+     * @throws IOException
+     */
+    private Query initQuery(GranuleSource granules)
+            throws TransformException, FactoryException, IOException {
+        List<Filter> filters = new ArrayList<Filter>();
+
+        Query query = Query.ALL;
+
+        // Set bbox query if a ROI has been provided
+        if (roiManager != null) {
+            CoordinateReferenceSystem targetCRS = roiManager.getTargetCRS();
+            GeometryDescriptor geomDescriptor = granules.getSchema().getGeometryDescriptor();
+            CoordinateReferenceSystem indexCRS = geomDescriptor.getCoordinateReferenceSystem();
+            ReferencedEnvelope envelope = null;
+            if (targetCRS != null && !roiManager.isRoiCrsEqualsTargetCrs()) {
+                envelope = new ReferencedEnvelope(
+                    roiManager.getSafeRoiInTargetCRS().getEnvelopeInternal(), targetCRS);
+                MathTransform reprojectionTrasform = null;
+                if (!CRS.equalsIgnoreMetadata(targetCRS, indexCRS)) {
+                    reprojectionTrasform = CRS.findMathTransform(targetCRS, indexCRS, true);
+                    if (!reprojectionTrasform.isIdentity()) {
+                        envelope = envelope.transform(indexCRS, true);
+                    }
+                }
+            } else {
+                envelope = new ReferencedEnvelope(roiManager.getSafeRoiInNativeCRS().getEnvelopeInternal(), indexCRS);
+            }
+
+            final PropertyName geometryProperty = FeatureUtilities.DEFAULT_FILTER_FACTORY
+                    .property(geomDescriptor.getName());
+            filters.add(FeatureUtilities.DEFAULT_FILTER_FACTORY.bbox(geometryProperty, envelope));
+        }
+
+        // Add the filter if specified
+        if (filter != null) {
+            filters.add(filter);
+        }
+
+        // Set the query filter
+        if (!filters.isEmpty()) {
+            query = new Query();
+            query.setFilter(FeatureUtilities.DEFAULT_FILTER_FACTORY.and(filters));
```

Comment:
An "and of one" is probably working but looks odd.  Since you have a list of filters, this method will do the right thing for you (you could dodge the if completely): https://github.com/geoserver/geoserver/blob/bab3171ef936628d28d6cbd0377a58f648716e9c/src/main/src/main/java/org/geoserver/catalog/Predicates.java#L240

Reply:
Cool. Thanks for the suggestion. going to update